sigs/jvm/rfcs/20190904-java-tensors.md

```diff
@@ -0,0 +1,301 @@
+higher rank. For example:
+
+```java
+public final class Constant<U> extends PrimitiveOp {
+
+    public static final Constant<TInt32> create(Scope scope, Integer value) {
+        ...
+    }
+    
+    public static final Constant<TInt32> create(Scope scope, Integer... values) {
+        ...
+    }
+    
+    public static final <U extends Tensor<?>> Constant<U> create(Scope scope, U data) {
+        ...
+    }
+}
+
+Constant<TInt32> scalar = tf.constant(10);  // single argument constructor is chosen
+
+Constant<TInt32> vector = tf.constant(10, 11, 42);  // variadic constructor is chosen
+
+Constant<TInt32> matrix = tf.constant(TInt32.ofShape(2, 2).row(10, 11).row(30, 40).done());
+
+```
+
+## Addendum
+
+### NdArray Iterative Data Initialization
+
+In some previous examples, some tensors are initialize using a new method called `row`. This method is intented
+to be added to the original [`NdArray` API](https://github.com/karllessard/community/blob/master/sigs/jvm/rfcs/20190606-java-tensor-io.md#ndarray-api)
+submitted in the previous RFC, with the objective of allowing the allocation and initialization
+of a small tensor in a single line to mimics standard multidimensional Java arrays.
+
+The idea is to initialize the data of each row of the `NdArray` sequentially by invoking `row(...)`, 
+and then invoke the `done()` method to return the newly allocated `NdArray` instance. A "row" in this terminology
+means a vector on the last axis (dimension) of an array.
+
+Let's compare standard Java arrays with those new methods as they are used with tensors:
+```java
+
+// Create a 2D matrix
+
+int[][] matrix2d_s = new int[][] { {1, 2}, {3, 4} };
+
+TInt32 matrix2d_t = TInt32.ofShape(2, 2).row(1, 2).row(3, 4).done();
+
+// Create a 3D matrix
+
+int[][][] matrix3d_s = new int[][][] { { {1, 2}, {3, 4} }, { {5, 6}, {7, 8} }, { {9, 10}, {11, 12} } };
+
+TInt32 matrix3d_t = TInt32.ofShape(3, 2, 2)
+  .row(1, 2).row(3, 4)
+  .row(5, 6).row(7, 8)
+  .row(9, 10).row(11, 12)
+  .done();
+```
+Of course, the standard Java arrays gives a better view on where is located a given row, by embedding it
+through a series of brackets. The actual proposal is focusing on simplicity and performance more that readability.
+*Please, feel free to suggest any better solution.*
+
+Note also that an advantage of using `NdArray` is that the type and the shape of the tensor is explicit and can be easily 
+retrieved by calling `t.dataType()` and `t.shape()` respectively, while with standard Java arrays those need
+to be inferred and discovered by the TF Java client using costly reflections and navigating through the array.
+
+### Custom Tensor Types
+
+It has been reported that in some cases, supporting compile-time type safety can become a nightmare when 
+testing a network using different data types. Actually, this can require a lot of search & replace on each 
+experiment (e.g. `FLOAT` vs `DOUBLE`). 
```

Comment:
If we expanded the set of input and output methods (i.e. so Tensor<type> had methods that accepted and returned all Java types and did the expected conversions) we could use a generic form of this, store the required enum or DataType implementation in a field and then make all tensor construction either accept that field or infer the type from other tensors. That would make it a single line change to change data types.

Reply:
Yeah I'm wondering if the Tensor should accept all Java types and handle the conversion... right now, it does expose all input/output methods but fail at runtime if the type is not compatible the tensor data type, like [here](https://github.com/tensorflow/tensorflow/blob/36e4ce080e77aa1126a3d26f22a4e7bd656a7c91/tensorflow/java/src/main/java/org/tensorflow/Tensor.java#L469).

There might be other issues as well, think of "UInt32"... how do you pass such a value in a 32-bit Integer?? It sounds like we might need to some conversion in the end...

Reply:
The problem with accepting all types is when we have value types and can more easily express uint32 it'll be irritating to deal with it. But the type system will be quite different then, and we're unlikely to upgrade to that Java version for a while after release (plus it's not ready yet).

Thinking more about this, maybe there should be a static factory on Tensor that accepts a DataType and an NDArray or other mass type and performs the appropriate conversions. That way the only thing that needs to change is the data type which can be a variable stored in a single place. That's pretty close to where we started off, but allowing it to make type conversions rather than just checking for type equality.

Reply:
I still don’t have a solution for supporting those kind of DataTypes... again, if we let the user pick the representation of his choice for initializing the data instead of forcing him to a specific one might help a bit (e.g. a user could decide to use `int`s, `float`s or `double`s or even `BigInteger/Decimal` to feed a `TFloat` tensor, instead of forcing him to use `floats`s only). We’ll then need a converter to map the provided values to the physical buffer... 

That’s something that wasn’t foreseen though and we should work on it if we all think that’s the right approach.